modules/userdirectory-canvas/src/main/java/org/opencastproject/userdirectory/canvas/CanvasUserProviderInstance.java

```diff
@@ -0,0 +1,705 @@
+        connection.setRequestMethod("GET");
+        connection.setDoOutput(true);
+        connection.setRequestProperty("Authorization", "Bearer " + canvasToken);
+        connection.setRequestProperty("User-Agent", OC_USERAGENT);
+
+        String json = IOUtils.toString(new BufferedInputStream(connection.getInputStream()));
+        logger.debug(json);
+
+        logger.debug("Response code for API: {}, is: {}", url.toString(), connection.getResponseCode());
+        JsonArray courses = new JsonParser().parse(json).getAsJsonArray();
+        for (JsonElement courseElement : courses) {
+          JsonObject course = courseElement.getAsJsonObject();
+          JsonElement jsonCourseId = course.get(courseIdentifierProperty);
+          if (jsonCourseId.isJsonNull()) {
+            continue;
+          }
+          String courseId = jsonCourseId.getAsString();
+          if (StringUtils.isNotBlank(courseId)) {
+            JsonArray enrollments = course.get("enrollments").getAsJsonArray();
+            for (JsonElement enrollmentElement : enrollments) {
+              JsonObject enrollment = enrollmentElement.getAsJsonObject();
+              /*
+              Canvas enrollment `type` is a parent/base type that roles are created from.
+              We allow the specifying of parent types and/or a child roles in configuration
+              (these are mutually exclusive in Canvas, you can't have a child role with the
+              same name as any parent type).
+              This provides the ability to use any parent type to grant the LTI
+              instructor roles for any child role created from it now or in the future, and/or
+              specific set of child roles.
+              There can be multiple enrollments for a user for a given course, adding them all to
+              the roleList is ok because the roleList is converted to a Set later on.
+              */
+              String type = enrollment.get("type").getAsString();
+              String role = enrollment.get("role").getAsString();
+              String opencastRoleFromCanvasType = buildOpencastRole(courseId, type);
+              roleList.add(opencastRoleFromCanvasType);
+              String opencastRoleFromCanvasRole = buildOpencastRole(courseId, role);
+              roleList.add(opencastRoleFromCanvasRole);
+            }
+          }
+        }
+
+        String linkHeader = connection.getHeaderField("link");
+        Matcher matcher = pattern.matcher(linkHeader);
+        if (matcher.find()) {
+          logger.debug("Link header contains a next page so there are more courses to fetch: {}", linkHeader);
+          nextPage = matcher.group(1);
+        } else {
+          logger.debug("No more courses to find for user: {}", userId);
+          nextPage = null;
+        }
+      }
+
+      if (roleList.isEmpty()) {
+        return null;
+      }
+      return roleList.toArray(new String[0]);
+
+    } catch (FileNotFoundException fnf) {
```

Comment:
Not sure if it's necessary but do we need to

- disconnect the connection?
- close the input stream?